client/service-container-chain/src/data_preservers.rs

```diff
@@ -0,0 +1,420 @@
+        /// so before calling `spawn` make sure to call `wait_for_paritydb_lock` before, like we do in
+        /// `handle_update_assignment`.
+        fn spawn(
+            &self,
+            container_chain_para_id: ParaId,
+            start_collation: bool,
+        ) -> impl std::future::Future<Output = ()> + Send {
+            let mut set = self.state.lock().unwrap();
+
+            let event = SpawnerEvent::Started(container_chain_para_id, start_collation);
+
+            assert!(set.remove(&event), "Unexpected event {event:?}");
+
+            async {}
+        }
+
+        /// Stop a container chain. Prints a warning if the container chain was not running.
+        /// Returns the database path for the container chain, can be used with `wait_for_paritydb_lock`
+        /// to ensure that the container chain has fully stopped. The database path can be `None` if the
+        /// chain was not running.
+        fn stop(&self, container_chain_para_id: ParaId, keep_db: bool) -> Option<PathBuf> {
+            let mut set = self.state.lock().unwrap();
+
+            let event = SpawnerEvent::Stopped(container_chain_para_id, keep_db);
+
+            assert!(set.remove(&event), "Unexpected event {event:?}");
+
+            None
+        }
+    }
+
+    #[tokio::test]
+    async fn task_logic_works() {
+        let spawner = MockSpawner::new();
+
+        let profile_id = 0;
+        let para_id1 = ParaId::from(1);
+        let para_id2 = ParaId::from(2);
+
+        tokio::spawn(task_watch_assignment(spawner.clone(), profile_id));
+        // Wait for task to start and subscribe to block stream.
+        tokio::time::sleep(Duration::from_millis(100)).await;
+
+        spawner.set_expectations(vec![SpawnerEvent::Started(para_id1, false)]);
+        spawner.chain_interface.mock_block({
+            let mut map = BTreeMap::new();
+            map.insert(profile_id, DataPreserverAssignment::Active(para_id1));
+            map
+        });
+        tokio::time::sleep(Duration::from_millis(100)).await;
+        spawner.ensure_all_events_were_emitted();
+
+        spawner.set_expectations(vec![SpawnerEvent::Stopped(para_id1, false)]);
+        spawner.chain_interface.mock_block({
+            let mut map = BTreeMap::new();
+            map.insert(profile_id, DataPreserverAssignment::NotAssigned);
+            map
+        });
+        tokio::time::sleep(Duration::from_millis(100)).await;
+        spawner.ensure_all_events_were_emitted();
+
+        spawner.set_expectations(vec![SpawnerEvent::Started(para_id2, false)]);
+        spawner.chain_interface.mock_block({
+            let mut map = BTreeMap::new();
+            map.insert(profile_id, DataPreserverAssignment::Active(para_id2));
+            map
+        });
+        tokio::time::sleep(Duration::from_millis(100)).await;
+        spawner.ensure_all_events_were_emitted();
+
+        spawner.set_expectations(vec![SpawnerEvent::Stopped(para_id2, true)]);
```

Comment:
It feels weird to have the expected events before the tests, in most other tests we put them in the final assert. But you can ignore this.

Reply:
Instead each function could collect the events then expect will ensure it matches :)

Reply:
disregard